src/Transform/OptimizeMemoryPools.cpp

```diff
@@ -0,0 +1,905 @@
+
+    // Construct the list of Values on which the current AllocOp depends on.
+    llvm::SetVector<Operation *> dependentOps;
+    while (operandList.size() > 0 && refsUseIsDisjoint) {
+      Value currentElement = operandList[0];
+      Operation *definingOperation = currentElement.getDefiningOp();
+
+      // If this value has not been seen before, process it.
+      if (definingOperation && dependentOps.count(definingOperation) == 0) {
+        // Add value to dependent values list.
+        dependentOps.insert(definingOperation);
+
+        if (llvm::dyn_cast<AffineLoadOp>(definingOperation) ||
+            llvm::dyn_cast<LoadOp>(definingOperation)) {
+          // Check that the MemRef operand of this load operation is
+          // not in the firstGetRefList.
+          Value loadOperand = definingOperation->getOperands()[0];
+          if (!isBlockArgument(definingOperation, loadOperand)) {
+            Operation *loadOperandDefinition = loadOperand.getDefiningOp();
+            KrnlGetRefOp loadGetRefOperand =
+                llvm::dyn_cast_or_null<KrnlGetRefOp>(loadOperandDefinition);
+
+            // If the load operand is valid, compare it with all the entries
+            // in the firstGetRefList. If it matches any one of them then the
+            // secondGetRef cannot share the same memory pool slot with the
+            // rest of the getref operations in the firstGetRefList.
+            if (loadGetRefOperand)
+              for (auto firstGetRef : firstGetRefList)
+                if (firstGetRef == loadGetRefOperand) {
+                  refsUseIsDisjoint = false;
+                  break;
+                }
+          }
+        } else {
+          // Add operands to work queue.
+          for (const auto &operand : definingOperation->getOperands())
+            if (!isBlockArgument(secondGetRef, operand))
+              operandList.emplace_back(operand);
+        }
+      }
+
+      // Erase first element from work queue.
+      operandList.erase(operandList.begin());
+    }
+
+    // Exit if use is not disjoint.
+    if (!refsUseIsDisjoint)
+      break;
+  }
+
+  return refsUseIsDisjoint;
+}
+
+/// secondGetRef candidate is checked against every element of firstGetRefList
+/// whether a load/store chain exists between them:
+///
+///    a = load firstGetRef[]
+///    b = f(a)
+///    store b secondGetRef[]
+///
+/// The check must be done both ways:
+///
+///    a = load secondGetRef[]
+///    b = f(a)
+///    store b firstGetRef[]
+///
+bool getRefUsesAreMutuallyDisjoint(SmallVector<KrnlGetRefOp, 4> firstGetRefList,
+    SmallVector<KrnlGetRefOp, 4> secondGetRefList) {
+  for (auto getRef : secondGetRefList) {
+    if (!getRefUsesAreDisjoint(firstGetRefList, getRef)) {
+      return false;
+    }
+  }
+
+  for (auto getRef : firstGetRefList) {
+    if (!getRefUsesAreDisjoint(secondGetRefList, getRef)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+/// Op is a LoadOp or AffineLoadOp.
```

Comment:
I am surprised that there is no query like this in MLIR. A cursory search could not find one. I wonder if we are not doing ourselves a disservice by burying these code deep into a .cpp. I would suggest to do a src/Support directory with a couple of system-wide useful queries like these. What do you think?

Reply:
Done